### PR TITLE
Add an analyzer that reports a diagnostic when MSBuildWorkspace likely needs upgrading

### PR DESCRIPTION
**Pull request by DustinCampbell**
This analyzer will report a diagnostic under the following conditions:

1. The current compilation references `Microsoft.CodeAnalysis.Workspaces.Desktop`.
2. The current compilation does not reference `Microsoft.CodeAnalysis.Workspaces.MSBuild`.
3. The current compilation does not contain the type `Microsoft.CodeAnalysis.Workspaces.MSBuildWorkspace`.
4. An unbound identifier is encountered with the name `MSBuildWorkspace`.

Ultimately, the diagnostic will simply point to guidance that helps customers use `MSBuildWorkspace` properly.

There is a still a bit of work yet before this can be merged:

- [x] Word smith the title, message, and description
- [x] Provide a link to usage guidance for MSBuildWorkspace

Finally, note that the unit tests do not verify the first two checks above. This is actually very hard to do with the current test infrastructure.

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Analyzers/CSharp/CSharpUpgradeMSBuildWorkspaceAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            context.RegisterSyntaxNodeAction(AnalyzerIdentifier, SyntaxKind.IdentifierName);
+        }
+
+        protected override void RegisterIdentifierAnalysis(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzerIdentifier, SyntaxKind.IdentifierName);
+        }
+
+        private void AnalyzerIdentifier(SyntaxNodeAnalysisContext context)
```
Comment (jasonmalinowski) [inline src/Microsoft.CodeAnalysis.Analyzers/CSharp/CSharpUpgradeMSBuildWorkspaceAnalyzer.cs]: Did you mean to call it AnalyzeIdentifier?

File: src/Microsoft.CodeAnalysis.Analyzers/VisualBasic/VisualBasicUpgradeMSBuildWorkspaceAnalyzer.vb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        End Sub
+
+        Private Sub AnalyzerIdentifier(context As SyntaxNodeAnalysisContext)
+            If TypeOf context.Node IsNot IdentifierNameSyntax Then
+                Return
+            End If
+
+            Dim identifierName = DirectCast(context.Node, IdentifierNameSyntax)
+
+            If identifierName.Identifier.ToString() <> MSBuildWorkspace Then
```
Comment (jasonmalinowski) [inline src/Microsoft.CodeAnalysis.Analyzers/VisualBasic/VisualBasicUpgradeMSBuildWorkspaceAnalyzer.vb]: I'm torn between pointing out this should be VB case insensitive, and just saying that those who do that deserve their own fate.

File: src/Microsoft.CodeAnalysis.Analyzers/VisualBasic/VisualBasicUpgradeMSBuildWorkspaceAnalyzer.vb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        Protected Overrides Sub RegisterIdentifierAnalysis(context As AnalysisContext)
+            context.RegisterSyntaxNodeAction(AddressOf AnalyzerIdentifier, SyntaxKind.IdentifierName)
+        End Sub
+
+        Private Sub AnalyzerIdentifier(context As SyntaxNodeAnalysisContext)
+            If TypeOf context.Node IsNot IdentifierNameSyntax Then
+                Return
+            End If
+
+            Dim identifierName = DirectCast(context.Node, IdentifierNameSyntax)
```
Comment (jasonmalinowski) [inline src/Microsoft.CodeAnalysis.Analyzers/VisualBasic/VisualBasicUpgradeMSBuildWorkspaceAnalyzer.vb]: TryCast?

**Comment by jasonmalinowski** [comment]
The one thing I *don't* know here is how this all ships. @mavasani do we chain these as a dependency of our NuGet packages somehow today?

**Comment by mavasani** [comment]
We need to fix https://github.com/dotnet/roslyn/issues/24899 once we have a signed build with this change.

File: src/Microsoft.CodeAnalysis.Analyzers/VisualBasic/VisualBasicUpgradeMSBuildWorkspaceAnalyzer.vb (hunk 2 above)
Comment (DustinCampbell) [reply src/Microsoft.CodeAnalysis.Analyzers/VisualBasic/VisualBasicUpgradeMSBuildWorkspaceAnalyzer.vb]: yeah, of course it should be. Will fix.

File: src/Microsoft.CodeAnalysis.Analyzers/VisualBasic/VisualBasicUpgradeMSBuildWorkspaceAnalyzer.vb (hunk 3 above)
Comment (DustinCampbell) [reply src/Microsoft.CodeAnalysis.Analyzers/VisualBasic/VisualBasicUpgradeMSBuildWorkspaceAnalyzer.vb]: sure

File: src/Microsoft.CodeAnalysis.Analyzers/CSharp/CSharpUpgradeMSBuildWorkspaceAnalyzer.cs (hunk 1 above)
Comment (DustinCampbell) [reply src/Microsoft.CodeAnalysis.Analyzers/CSharp/CSharpUpgradeMSBuildWorkspaceAnalyzer.cs]: oops :smile:

**Comment by mavasani** [comment]
@DustinCampbell good to merge?

**Comment by DustinCampbell** [comment]
> @DustinCampbell good to merge?

No. Please read PR description. :smile:

I need to actually put something useful into the message before this is a valuable diagnostic.

**Comment by mavasani** [comment]
Ah yes, sure. Thanks.

**Comment by DustinCampbell** [comment]
Ooops! Let me update the loc files.

**Pull request merged**